ZenPacks/zenoss/ZenPackLib/lib/base/ZenPack.py

```diff
@@ -190,25 +149,129 @@ def remove(self, app, leaveObjects=False):
+                else:
+                    pass
+
+    def check_diff(self, app, parent, relname, object, spec, specparam):
+        """Return True if object has changed creating preupgrade backup if needed"""
+        diff = self.object_changed(app, object, spec, specparam)
+        if diff:
+            # preserve the existing object if different
+            time_str = time.strftime("%Y%m%d%H%M", time.localtime())
+            preupgrade_id = "{}-preupgrade-{}".format(object.id, time_str)
+            self.move_object(parent, relname, object.id, preupgrade_id)
+            LOG.info("Existing object {}/{} differs from "
+                     "the newer version included with the {} ZenPack.  "
+                     "The existing object will be "
+                     "backed up to '{}'.  Please review and reconcile any "
+                     "local changes before deleting the backup:\n{}".format(
+                        parent.getDmdKey(), object.id, self.id, preupgrade_id, diff))
+            return True
+        return False
+
+    def get_object(self, parent, relname, object_id):
+        """Attempt to retrieve an object given its id, parent instance, and relation name"""
+        rel = getattr(parent, relname, None)
+        if rel:
+            try:
+                return rel._getOb(object_id)
+            except AttributeError:
+                pass
+        return None
+
+    def get_or_create_backup(self, parent, relname, object_id, overwrite=True):
+        """Create and return a backup of object"""
+        backup_id = "{}-backup".format(object_id)
+        backup_target = self.get_object(parent, relname, backup_id)
+        # return or delete the template if it already exists
+        # which could occur if zenpack installation fails and is re-attempted
+        if backup_target:
+            if overwrite:
+                backup_target.getPrimaryParent()._delObject(backup_target.id)
+            else:
+                return backup_target
+        # move the object to its backup
+        self.rename_object(parent, relname, object_id, backup_id)
+        return self.get_object(parent, relname, backup_id)
+
+    def rename_object(self, parent, relname, source_id, dest_id):
+        """execute manage_renameObject and log exceptions"""
+        rel = getattr(parent, relname, None)
+        if rel:
+            try:
+                rel.manage_renameObject(source_id, dest_id)
+            except Exception as e:
+                LOG.warn("Could not move template {}/{} to {}/{} ({})".format(parent.getDmdKey(), source_id,
+                                                                              parent.getDmdKey(), dest_id,
+                                                                              e))
+        else:
+            LOG.warn("Cannot rename {} on nonexistent relation {}".format(source_id, relname))
+
+    def move_object(self, parent, relname, source_id, dest_id, overwrite=True):
```

Comment:
I don't see the overwrite argument actually being used. I'd stick to the [YAGN principle](https://en.wikipedia.org/wiki/You_aren't_gonna_need_it) here and not create code paths that aren't used.